### PR TITLE
Limit Concurrent DNS queries for ClusterConnectionManager and DnsMonitor

### PR DESCRIPTION
**Pull request by umutkocasarac**

### CODE REVIEW

**Comment by mrniko** [comment]
Thanks for proposed changes. But this issue should be resolved with Netty's DNS cache.

**Comment by umutkocasarac** [comment]
Thank you @mrniko just for clarification, it is not netty DNS cache issue. Because of our big redis cluster, cluster scaner is making ~200 DNS queries concurrently and it is creating a lot of pressure on netty dns resolver side and sometimes we are getting a timeout. I also think it should be solved on netty, however, it is also important to have control over the netty and not creating a lot of pressure. 

Based on this information, I want you to reconsider this PR.

**Comment by mrniko** [comment]
If DNS cache works properly then why you get 200+ DNS queries?

**Comment by umutkocasarac** [comment]
because we have 200+ redis nodes on one cluster

**Comment by mrniko** [comment]
Isn't DNS cache should return cached result for each Redis node?

**Comment by umutkocasarac** [comment]
It returns cached results if it has, however, most of the time it doesn't have any cache because redis DNS TTL is short and almost all of them expire at the same time. I can manually increase the TTL parameter on the netty DNS cache but it will only reduce the frequency of the error.

**Comment by syllabix** [comment]
hello @mrniko - this is indeed an issue we are seeing with large clusters as well. The primary issue is that the DNS cache _does_ return the cached result up until the TTL from the server expires (for example 15 seconds). As a result, when an application first starts up, it has no records in cache, and thus needs to make a large amount of DNS queries on app start (which can effect readiness, and apps will not start). In the event they do start - every TTL interval, the entries expire and need to be refreshed, and as such a significant volume of requests are issued once again, re-surfacing the problem with timeouts. the approach @umutkocasarac is suggested batches these queries in such away that it not all records are fetched at once, mitigating resource contention/timeouts for large clusters. 

considering this is case that is probably not common to most users of Redisson, keeping this feature off by default might can help support users with larger clusters in the event they need it.

File: redisson/src/main/java/org/redisson/config/BaseMasterSlaveServersConfig.java [hunk 1]
```diff
@@ -314,5 +316,22 @@ public T setDnsMonitoringInterval(long dnsMonitoringInterval) {
...
+
+    /**
+     * Max concurrent calls for the slave dns monitoring. It is disabled by default
+     * Set <code>0</code> to disable.
+     * @param maxConcurrentDnsQuery max concurrent dns call
+     * @return config
+     */
+    public T setMaxConcurrentDnsQuery(int maxConcurrentDnsQuery) {
+        if (maxConcurrentDnsQuery < 0) {
+            throw new IllegalArgumentException("maxConcurrentDnsQuery can be lower than 0");
```
Comment (SplotyCode) [inline redisson/src/main/java/org/redisson/config/BaseMasterSlaveServersConfig.java:328]: Do you mean 'can not'?

Comment (umutkocasarac) [reply redisson/src/main/java/org/redisson/config/BaseMasterSlaveServersConfig.java]: yes, fixed the message

**Comment by mrniko** [comment]
@syllabix From javadocs
`If the TTL of the DNS resource record returned by the DNS server is less than the minimum TTL or greater than the maximum TTL, this resolver will ignore the TTL from the DNS server and use the minimum TTL or the maximum TTL instead respectively.
The default value is {@code 0} and {@link Integer#MAX_VALUE}, which practically tells this resolver to respect the TTL from the DNS server.` How did you define DnsCache object?

**Comment by mrniko** [comment]
@syllabix Did you try to define `.ttl(60*60*2, 60*60*2)` ?

**Comment by mrniko** [comment]
Can you try LimitedDnsAddressResolverGroupFactory below?

```java
    public static class LimitedDnsAddressResolverGroupFactory implements AddressResolverGroupFactory {

        private final ThreadPoolExecutor ee;
        private final int minttl;
        private final int maxttl;

        public LimitedDnsAddressResolverGroupFactory(int minttl, int maxttl, int maxConcurrentQueries, Duration delayBetweenQueries) {
            this.minttl = minttl;
            this.maxttl = maxttl;
            ee = new ThreadPoolExecutor(1, 1,
                    0L, TimeUnit.MILLISECONDS,
                    new LinkedBlockingQueue<>()) {

                final AtomicLong s = new AtomicLong();

                @Override
                protected void beforeExecute(Thread t, Runnable r) {
                    if (s.incrementAndGet() % maxConcurrentQueries == 0) {
                        try {
                            t.join(delayBetweenQueries.toMillis());
                        } catch (InterruptedException e) {
                            throw new RuntimeException(e);
                        }
                    }
                }
            };
        }

        @Override
        public AddressResolverGroup<InetSocketAddress> create(Class<? extends DatagramChannel> channelType, DnsServerAddressStreamProvider nameServerProvider) {
            final DnsNameResolverBuilder builder = new DnsNameResolverBuilder();
            builder.channelType(NioDatagramChannel.class)
                    .nameServerProvider(DnsServerAddressStreamProviders.platformDefault())
                    .ttl(minttl, maxttl);

            final DnsAddressResolverGroup group = new DnsAddressResolverGroup(builder) {
                @Override
                protected AddressResolver<InetSocketAddress> newAddressResolver(EventLoop eventLoop, NameResolver<InetAddress> resolver) throws Exception {
                    return new InetSocketAddressResolver(eventLoop, resolver) {

                        @Override
                        protected void doResolve(InetSocketAddress unresolvedAddress, Promise<InetSocketAddress> promise) throws Exception {
                            ee.submit(() -> {
                                resolver.resolve(unresolvedAddress.getHostName())
                                        .addListener(new FutureListener<InetAddress>() {
                                            @Override
                                            public void operationComplete(Future<InetAddress> future) throws Exception {
                                                if (future.isSuccess()) {
                                                    promise.setSuccess(new InetSocketAddress(future.getNow(), unresolvedAddress.getPort()));
                                                } else {
                                                    promise.setFailure(future.cause());
                                                }
                                            }
                                        });
                            });
                        }

                        @Override
                        protected void doResolveAll(InetSocketAddress unresolvedAddress, Promise<List<InetSocketAddress>> promise) throws Exception {
                            ee.submit(() -> {
                                resolver.resolveAll(unresolvedAddress.getHostName())
                                        .addListener(new FutureListener<List<InetAddress>>() {
                                            @Override
                                            public void operationComplete(Future<List<InetAddress>> future) throws Exception {
                                                if (future.isSuccess()) {
                                                    List<InetAddress> inetAddresses = future.getNow();
                                                    List<InetSocketAddress> socketAddresses =
                                                            new ArrayList<InetSocketAddress>(inetAddresses.size());
                                                    for (InetAddress inetAddress : inetAddresses) {
                                                        socketAddresses.add(new InetSocketAddress(inetAddress, unresolvedAddress.getPort()));
                                                    }
                                                    promise.setSuccess(socketAddresses);
                                                } else {
                                                    promise.setFailure(future.cause());
                                                }
                                            }
                                        });
                            });
                        }
                    };
                }
            };
            return group;
        }
    }

```

Each resolve method delayed by a second.

**Comment by mrniko** [comment]
I updated the code in my comment above

**Comment by mrniko** [comment]
Updated my code above once again.

**Comment by umutkocasarac** [comment]
> I updated the code in my comment above

I have just deployed the code, and I don't see any error for the last 15 minutes, will update the thread tomorrow morning with the result

**Comment by mrniko** [comment]
@umutkocasarac 

Which rate limit params you specify in constructor?

**Comment by umutkocasarac** [comment]
you can find the configuration at below @mrniko 
```
public LimitedDnsAddressResolverGroupFactory(){
        this(10, Duration.ofMillis(10));
    }
```

**Comment by mrniko** [comment]
Is it AWS Elasticache hosting?

**Comment by mrniko** [comment]
@umutkocasarac 

Here is another solution. Which is more optimal in my opinion.

```java
import io.netty.channel.EventLoop;
import io.netty.channel.socket.DatagramChannel;
import io.netty.resolver.AddressResolver;
import io.netty.resolver.AddressResolverGroup;
import io.netty.resolver.InetSocketAddressResolver;
import io.netty.resolver.NameResolver;
import io.netty.resolver.dns.DnsAddressResolverGroup;
import io.netty.resolver.dns.DnsServerAddressStreamProvider;
import io.netty.util.concurrent.Future;
import io.netty.util.concurrent.FutureListener;
import io.netty.util.concurrent.Promise;

import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;

/**
 *
 * @author Nikita Koksharov
 *
 */
public class SequentalDnsAddressResolverFactory implements AddressResolverGroupFactory {

    private final ExecutorService ee;

    public SequentalDnsAddressResolverFactory(int threads) {

        ee = Executors.newFixedThreadPool(threads);
    }

    @Override
    public AddressResolverGroup<InetSocketAddress> create(Class<? extends DatagramChannel> channelType, DnsServerAddressStreamProvider nameServerProvider) {
        DnsAddressResolverGroup group = new DnsAddressResolverGroup(channelType, nameServerProvider) {
            @Override
            protected AddressResolver<InetSocketAddress> newAddressResolver(EventLoop eventLoop, NameResolver<InetAddress> resolver) throws Exception {
                return new InetSocketAddressResolver(eventLoop, resolver) {

                    @Override
                    protected void doResolve(InetSocketAddress unresolvedAddress, Promise<InetSocketAddress> promise) throws Exception {
                        ee.submit(() -> {
                            resolver.resolve(unresolvedAddress.getHostName())
                                    .addListener(new FutureListener<InetAddress>() {
                                        @Override
                                        public void operationComplete(Future<InetAddress> future) throws Exception {
                                            if (future.isSuccess()) {
                                                promise.setSuccess(new InetSocketAddress(future.getNow(), unresolvedAddress.getPort()));
                                            } else {
                                                promise.setFailure(future.cause());
                                            }
                                        }
                                    });
                            try {
                                promise.await();
                            } catch (InterruptedException e) {
                                Thread.currentThread().interrupt();
                            }
                        });
                    }

                    @Override
                    protected void doResolveAll(InetSocketAddress unresolvedAddress, Promise<List<InetSocketAddress>> promise) throws Exception {
                        ee.submit(() -> {
                            resolver.resolveAll(unresolvedAddress.getHostName())
                                    .addListener(new FutureListener<List<InetAddress>>() {
                                        @Override
                                        public void operationComplete(Future<List<InetAddress>> future) throws Exception {
                                            if (future.isSuccess()) {
                                                List<InetAddress> inetAddresses = future.getNow();
                                                List<InetSocketAddress> socketAddresses =
                                                        new ArrayList<InetSocketAddress>(inetAddresses.size());
                                                for (InetAddress inetAddress : inetAddresses) {
                                                    socketAddresses.add(new InetSocketAddress(inetAddress, unresolvedAddress.getPort()));
                                                }
                                                promise.setSuccess(socketAddresses);
                                            } else {
                                                promise.setFailure(future.cause());
                                            }
                                        }
                                    });
                            try {
                                promise.await();
                            } catch (InterruptedException e) {
                                Thread.currentThread().interrupt();
                            }
                        });
                    }
                };
            }
        };
        return group;
    }
}
```
Can you try it with param: 2 threads up to 10? Let me know which is best for you.

**Comment by umutkocasarac** [comment]
Hi @mrniko,
I have tested your previous approach (LimitedDnsAddressResolverGroupFactory) for a day and it didn't get any error.
Thanks

**Comment by mrniko** [comment]
@umutkocasarac 

Can you also try another solution also? I consider to set it as default AddressResolverGroupFactory.

**Comment by umutkocasarac** [comment]
sure, but which one do you want me to test, DnsAddressResolverRateLimiterFactory or SequentialDnsAddressResolverFactory?

**Comment by mrniko** [comment]
@umutkocasarac 

SequentialDnsAddressResolverFactory with threads amount 2, 4, 6, 8, 10. Just want to know at which threads amount errors starts to appear for you case.

**Comment by Frank591** [comment]
Hello, @mrniko !
I am a colleague of Umut. I will do the tests with SequentialDnsAddressResolverFactory and will try to provide results tomorrow

**Comment by Frank591** [comment]
Hello, @mrniko .
Let me provide test results for `LimitedDnsAddressResolverGroupFactory` and `SequentialDnsAddressResolverFactory` .
### Test environment
**Elasticache cluster**
Node type: cache.t4g.micro
Shard configuration: 2 nodes per shard
Amount of shards: 96 shards
Elasticache version 6.2.6
Encryption in transit: Enabled
Multi-AZ: enabled

**Application** 
EC2 instance type: c6i.2xlarge
Amount of instances:3 
Redisson version: issue started with Redisson version 3.17.0. During the investigation it was updated up to 3.17.6
Java version:  java-17-amazon-corretto-jdk_17.0.3.6-1_amd64
OS version:  ubuntu-bionic-18.04-amd64
Netty version: 4.1.75
Redisson config: 
     fun createClusterConfig(
        nodeUris: List<String>,
        connectionPoolSize: Int,
        params: RedisConnectionParameters,
        masterReadProbability: Double
    ): Config {
        val config = Config().apply {
            nettyThreads = 1
            addressResolverGroupFactory = <Factory for test>
        }
        config.useClusterServers().apply {
            nodeUris.forEach {
                addNodeAddress(it)
            }
            connectTimeout = 30000
            timeout = 5000
            tcpNoDelay = true
            keepAlive = true
            dnsMonitoringInterval = 3000
            scanInterval = 3000
            masterConnectionMinimumIdleSize = 1
            masterConnectionPoolSize = 0
            slaveConnectionMinimumIdleSize = 1
            slaveConnectionPoolSize = 0
        }

        return config
    }

### Test results
**DnsAddressResolverRateLimiterFactory(250, Duration.ofSeconds(5))** 
Application didn't start at all, because of errors on initial connection to cluster during app start

**DnsAddressResolverRateLimiterFactory(10, Duration.ofMillis(10))** 
Application started succesfully on all servers. 
Test duration: 10 min
Amount of errors: 1 error case on 1 server

**SequentialDnsAddressResolverFactory(concurrency=2)**
The application started successfully on all servers. 
Test duration: 10 min
Amount of errors: no errors

**SequentialDnsAddressResolverFactory(concurrency=4)**
The application started successfully on all servers. 
Test duration: 90 min
Amount of errors: no errors

**SequentialDnsAddressResolverFactory(concurrency=6)**
The application started successfully on all servers. 
Test duration: 10 min
Amount of errors: no errors


**SequentialDnsAddressResolverFactory(concurrency=10)**
The application started successfully on all servers. 
Test duration: 120 min
Amount of errors: no errors detected

App version with` SequentialDnsAddressResolverFactory(concurrency=10)` will work this night. I will add updates tomorrow about this long test. If this test will be successful, I will continue tests on the production
Pls let me know if any additional information is required from our side

**Comment by Frank591** [comment]
@mrniko , small update about `SequentialDnsAddressResolverFactory(concurrency=10)` - unfortunately our test env was automatically stopped at night, so this version was tested only for ~7 hours. No errors were detected during the test

**Comment by mrniko** [comment]
@Frank591 

Thanks for testing! I'm closing this PR.

**Comment by Frank591** [comment]
@mrniko , thanks a lot for your help!
One more question from my side:
Do I understand correctly that for now same event loop is used for network operations to Redis and for DNS resolving? If yes, can it potentially be a problem? Does it make sense to introduce a separate event loop for DNS requests or not?

**Pull request closed without merging**